src/main/java/appeng/util/item/AESharedItemStack.java

```diff
@@ -0,0 +1,195 @@
+	{
+		return this.itemStack;
+	}
+
+	@Override
+	public int hashCode()
+	{
+		return Objects.hash( itemStack.getItem(), itemStack.getItemDamage() );
+	}
+
+	@Override
+	public boolean equals( final Object obj )
+	{
+		if( obj instanceof AESharedItemStack )
+		{
+			final AESharedItemStack other = (AESharedItemStack) obj;
+
+			if( this.itemStack == other.itemStack )
+			{
+				return true;
+			}
+			return ItemStack.areItemStacksEqual( this.itemStack, other.itemStack );
+		}
+		return false;
+	}
+
+	@Override
+	public int compareTo( AESharedItemStack b )
+	{
+		if( this.itemStack == b.itemStack )
+		{
+			return 0;
+		}
+
+		final int id = Item.getIdFromItem( this.itemStack.getItem() ) - Item.getIdFromItem( b.itemStack.getItem() );
+		if( id != 0 )
+		{
+			return id;
+		}
+
+		final int damageValue = this.itemStack.getItemDamage() - b.itemStack.getItemDamage();
+		if( damageValue != 0 )
+		{
+			return damageValue;
+		}
+
+		return this.compareNBT( b );
+	}
+
+	private int compareNBT( final AESharedItemStack b )
+	{
+		if( this.itemStack.getTagCompound() == b.itemStack.getTagCompound() )
+		{
+			return 0;
+		}
+
+		if( this.itemStack.getTagCompound() == LOW_TAG )
+		{
+			return -1;
+		}
+		if( this.itemStack.getTagCompound() == HIGH_TAG )
+		{
+			return 1;
+		}
+
+		final int nbt = ( this.itemStack == null ? 0 : this.itemStack.hashCode() ) - ( b.itemStack == null ? 0 : b.itemStack.hashCode() );
+		if( nbt != 0 )
+		{
+			return nbt;
+		}
+
+		return System.identityHashCode( this.itemStack.getTagCompound() ) - System.identityHashCode( b.itemStack.getTagCompound() );
+	}
+
+	AESharedItemStack getLow( final FuzzyMode fuzzy, final boolean ignoreMeta )
+	{
+		final ItemStack newDef = this.itemStack.copy();
+
+		if( ignoreMeta )
+		{
+			newDef.setItemDamage( 0 );
+		}
+		else
+		{
+			if( newDef.getItem().isDamageable() )
+			{
+				if( fuzzy == FuzzyMode.IGNORE_ALL )
+				{
+					newDef.setItemDamage( 0 );
+				}
+				else if( fuzzy == FuzzyMode.PERCENT_99 )
+				{
+					if( this.itemStack.getItemDamage() == 0 )
+					{
+						newDef.setItemDamage( 0 );
+					}
+					else
+					{
+						newDef.setItemDamage( 0 );
+					}
+				}
+				else
+				{
+					final int breakpoint = fuzzy.calculateBreakPoint( this.itemStack.getMaxDamage() );
+					newDef.setItemDamage( breakpoint <= this.itemStack.getItemDamage() ? breakpoint : 0 );
+				}
+			}
+			newDef.setTagCompound( LOW_TAG );
+		}
+
+		return new AESharedItemStack( newDef );
+	}
+
+	AESharedItemStack getHigh( final FuzzyMode fuzzy, final boolean ignoreMeta )
+	{
+		final ItemStack newDef = this.itemStack.copy();
+
+		if( ignoreMeta )
+		{
+			newDef.setItemDamage( Integer.MAX_VALUE );
+		}
+		else
+		{
+			if( newDef.getItem().isDamageable() )
+			{
+				if( fuzzy == FuzzyMode.IGNORE_ALL )
+				{
+					newDef.setItemDamage( this.itemStack.getMaxDamage() + 1 );
+				}
+				else if( fuzzy == FuzzyMode.PERCENT_99 )
+				{
+					if( this.itemStack.getItemDamage() == 0 )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)